serializers/src/main/java/io/pravega/schemaregistry/serializers/AbstractSerializer.java

```diff
@@ -0,0 +1,110 @@
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.contract.data.VersionInfo;
+import io.pravega.schemaregistry.schemas.Schema;
+import lombok.Getter;
+import lombok.SneakyThrows;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicReference;
+
+abstract class AbstractSerializer<T> extends BaseSerializer<T> {
+    private static final byte PROTOCOL = 0x1;
+
+    private final String groupId;
+    
+    private final SchemaInfo schemaInfo;
+    private final AtomicReference<EncodingId> encodingId;
+    private final boolean encodeHeader;
+    private final SchemaRegistryClient client;
+    @Getter
+    private final Codec codec;
+    private final boolean registerSchema;
+    
+    protected AbstractSerializer(String groupId,
+                                 SchemaRegistryClient client,
+                                 Schema<T> schema,
+                                 Codec codec,
+                                 boolean registerSchema, 
+                                 boolean encodeHeader) {
+        Preconditions.checkNotNull(groupId);
+        Preconditions.checkNotNull(client);
+        Preconditions.checkNotNull(codec);
+        Preconditions.checkNotNull(schema);
+        
+        this.groupId = groupId;
+        this.client = client;
+        this.schemaInfo = schema.getSchemaInfo();
+        this.registerSchema = registerSchema;
+        this.encodingId = new AtomicReference<>();
+        this.codec = codec;
+        this.encodeHeader = encodeHeader;
+        initialize();
+    }
+    
+    private void initialize() {
+        VersionInfo version;
+        if (registerSchema) {
+            // register schema
+            version = client.addSchema(groupId, schemaInfo);
+        } else {
+            // get already registered schema version. If schema is not registered, this will throw an exception. 
+            version = client.getVersionForSchema(groupId, schemaInfo);
+        }
+        if (encodeHeader) {
+            encodingId.set(client.getEncodingId(groupId, version, codec.getCodecType().getName()));
+        }
+    }
+    
+    @SneakyThrows(IOException.class)
+    @Override
+    public ByteBuffer serialize(T obj) {
+        ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
+        
+        // if schema is not null, pass the schema to the serializer implementation
+        if (schemaInfo != null) {
+            serialize(obj, schemaInfo, dataStream);
+        } else {
+            serialize(obj, null, dataStream);
+        }
+
+        dataStream.flush();
+
+        byte[] serialized = dataStream.toByteArray();
+        
+        ByteBuffer byteBuffer;
+        if (this.encodeHeader) {
+            Preconditions.checkNotNull(schemaInfo);
+            ByteBuffer encoded = codec.encode(ByteBuffer.wrap(serialized));
```

Comment:
passing ByteArrayOutputStream to codec enforces codec to actually copy the content from input source to bytearrayoutputstream.
If the codec does not wish to modify the content, it could simply return the received input source. like with NoOp codec.